### ISSUE TITLE
Cardinality of property '__tname__' of object type  is unknown when selecting in `with` scope inside `for` scope inside `with` scope

### ISSUE THREAD

**Issue by marcodlk** [edgedb/edgedb#5471]
This error has been raised in other issues (#4719, #5243), but I believe this is a distinct context.

- EdgeDB Version: 2.14+bc002ae
- EdgeDB CLI Version: EdgeDB CLI 2.3.1+5d93f42
- OS Version: Mac OS X Monterey

Steps to Reproduce:

The first edgeql command below works as expected, but simply moving the select into the top-level with causes the error.
```
EdgeDB 2.14+bc002ae (repl 2.3.1+5d93f42)
Type \help for help, \quit to quit.
edgedb> with
.......   data := {0, 1, 2},
....... select (
.......   for item in data union (
.......     with
.......       user := (select User limit 1)
.......     select user
.......   )
....... );
{
  default::User {id: 8e81e2ee-e8fa-11ed-bc44-7b514e5cfbe8},
  default::User {id: 8e81e2ee-e8fa-11ed-bc44-7b514e5cfbe8},
  default::User {id: 8e81e2ee-e8fa-11ed-bc44-7b514e5cfbe8},
}
edgedb> with
.......   data := {0, 1, 2},
.......   items := (
.......     for item in data union (
.......       with
.......         user := (select User limit 1)
.......       select user
.......     )
.......   )
....... select items;
edgedb error: InternalServerError: cardinality of property '__tname__' of object type 'default::User' is unknown
  Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
  Server traceback:
      Traceback (most recent call last):
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/server/compiler_pool/worker_proc.py", line 55, in worker
          res = meth(*args)
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/server/compiler_pool/worker.py", line 160, in compile
          units, cstate = COMPILER.compile(
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/server/compiler/compiler.py", line 2524, in compile
          unit_group = self._compile(ctx=ctx, source=source)
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/server/compiler/compiler.py", line 2049, in _compile
          return self._try_compile(ctx=ctx, source=source)
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/server/compiler/compiler.py", line 2113, in _try_compile
          comp, capabilities = self._compile_dispatch_ql(
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/server/compiler/compiler.py", line 2023, in _compile_dispatch_ql
          query = self._compile_ql_query(ctx, ql, script_info=script_info)
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/server/compiler/compiler.py", line 750, in _compile_ql_query
          out_type_data, out_type_id = sertypes.TypeSerializer.describe(
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/server/compiler/sertypes.py", line 582, in describe
          type_id = builder._describe_type(
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/server/compiler/sertypes.py", line 297, in _describe_type
          if ptr.singular(self.schema):
        File "/Users/delannmt/Library/Application Support/edgedb/portable/2.14/lib/python3.10/site-packages/edb/schema/pointers.py", line 751, in singular
          raise AssertionError(f'cardinality of {vn} is unknown')
      AssertionError: cardinality of property '__tname__' of object type 'default::User' is unknown
```

Providing my schema but note that the error happens regardless of the object being used.

```
    type User {
        required property name -> str {
            constraint exclusive;
        };
        required property hashed_password -> str;
        required property email -> str {
            constraint exclusive;
        };
        property is_admin -> bool {
            default := false;
        }
        property is_active -> bool {
            default := true;
        }
    }
```

**Comment by msullivan** [comment]
Fixed by #5508

**Comment by aljazerzen** [comment]
For reference, this is minimal reproducible example (using empty schema):

```
 with
   data := {0, 1, 2},
   items := (
     for item in data union (
       with
         user := (select schema::Object limit 1)
       select user
     )
   )
 select items;
```

`items` binding, for union and `user` binding are all required.
